extensions/amp-story/1.0/amp-story.js

```diff
@@ -348,6 +358,75 @@ export class AmpStory extends AMP.BaseElement {
   }
 
 
+  /** @private */
+  initializeStyles_() {
+    const storyId = this.getId();
+    const storyIdClassName = `i-amphtml-story-id-${storyId}`;
+    const pageEls = this.element.querySelectorAll('amp-story-page') || [];
+    const isStandalone = this.isStandalone_();
+    let styleEl;
+
+    if (isStandalone) {
+      styleEl = document.querySelector('style[amp-custom]');
+    } else {
+      styleEl = document.querySelector(
+          `style[amp-story="${escapeCssSelectorIdent(storyId)}"]`);
+    }
+
+    if (!styleEl) {
+      return;
+    }
+
+    this.mutateElement(() => {
+      styleEl.setAttribute('amp-story', storyId);
+      this.rewriteStyles_(styleEl);
+      Array.prototype.forEach.call(pageEls, pageEl => {
+        pageEl.classList.add(storyIdClassName);
+      });
+    });
+  }
+
+
+  /**
+   * @param {!Element} styleEl
+   * @private
+   */
+  rewriteStyles_(styleEl) {
+    if (!isExperimentOn(this.win, 'amp-story-responsive-units')) {
+      return;
+    }
+
+    // TODO(#15955): Update this to use CssContext from
+    // ../../../extensions/amp-animation/0.1/web-animations.js
+    styleEl.textContent = styleEl.textContent
+        .replace(/([\d.]+)vh/gmi, 'calc($1 * var(--i-amphtml-story-vh))')
+        .replace(/([\d.]+)vw/gmi, 'calc($1 * var(--i-amphtml-story-vw))')
+        .replace(/([\d.]+)vmin/gmi, 'calc($1 * var(--i-amphtml-story-vmin))')
+        .replace(/([\d.]+)vmax/gmi, 'calc($1 * var(--i-amphtml-story-vmax))');
+  }
+
+
+  /**
+   * @private
+   */
+  updateViewportSizeStyles_() {
+    this.vsync_.run({
+      measure: state => {
+        state.vh = this.activePage_.element.clientHeight / 100;
+        state.vw = this.activePage_.element.clientWidth / 100;
+        state.vmin = Math.min(state.vh, state.vw);
+        state.vmax = Math.max(state.vh, state.vw);
+      },
+      mutate: state => {
+        this.element.setAttribute('style',
```

Comment:
Using [setStyles](https://github.com/ampproject/amphtml/blob/master/src/style.js#L149) might be slightly better, for the edge case where a publisher would already by using the `style` attribute (if we even allow it?).

Reply:
setStyles doesn't work with CSS custom properties ☹️